src/api/lock.jl

```diff
@@ -0,0 +1,15 @@
+const liblock = ReentrantLock()
+
+# Try to acquire the lock (test-test-set) and close if successful
+# Otherwise, defer finalization
+# https://github.com/JuliaIO/HDF5.jl/issues/1048
+function try_close_finalizer(x)
+    if !islocked(liblock) && trylock(liblock) do
+        close(x)
+        true
+    end
+    else
+        finalizer(try_close_finalizer, x)
+    end
+    return nothing
+end
```

Comment:
It's a bit weird that this calls `close`, which hasn't yet been defined?

Reply:
`close` here is `Base.close`, so technically it has been defined.

Reply:
If we moved it,  where would we put it?

After the types are defined?

Reply:
Create a new file `common.jl` before `properties.jl`? That said, I don't feel too strongly about this.

Reply:
I frankly do not see the difference. Something will appear out of order any way we do this. If we put this in after types.jl, then we refer to `try_close_finalizer` before we define it.

Moreover I like it where it is. When someone goes to figure out see how the lock works, they'll see this weird piece of code nearby, which may be useful for future needs.